src/Middleware/Csrf.php

```diff
@@ -0,0 +1,115 @@
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Yiisoft\Security\Random;
+use Yiisoft\Security\TokenMasker;
+use Yiisoft\Yii\Web\Cookie;
+use Yiisoft\Yii\Web\Session\SessionInterface;
+
+final class Csrf implements MiddlewareInterface
+{
+    public const CSRF_HEADER = 'X-CSRF-Token';
+
+    private $name = '_csrf';
+    private $cookieParams = ['httpOnly' => true];
+    private $enableCookie = true;
+    private $responseFactory;
+    private $session;
+
+    private function __construct(ResponseFactoryInterface $responseFactory, SessionInterface $session)
+    {
+        $this->responseFactory = $responseFactory;
+        $this->session = $session;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $token = $this->getCsrfToken($request);
+
+        if (empty($token)) {
+            $token = Random::string();
+        }
+        $request = $request->withAttribute('csrf_token', TokenMasker::mask($token));
+
+        try {
+            $response = $handler->handle($request);
+            $response = $this->addTokenToResponse($response, $token);
+        } catch (\Throwable $e) {
+            throw new $e;
+        }
+
+        return $this->validateCsrfToken($request, $response);
+    }
+
+    private function addTokenToResponse(ResponseInterface $response, $token): ResponseInterface
+    {
+
+        if ($this->enableCookie) {
+            $cookieParameters = \array_merge($this->session->getCookieParameters(), $this->cookieParams);
+
+            $sessionCookie = (new Cookie($this->name, $token))
+                ->path($cookieParameters['path'])
+                ->domain($cookieParameters['domain'])
+                ->httpOnly($cookieParameters['httponly'])
+                ->secure($cookieParameters['secure'])
+                ->sameSite($cookieParameters['samesite']);
+
+            return $sessionCookie->addToResponse($response);
+        } else {
+            $this->session->set($this->name, $token);
+            return $response;
+        }
+    }
+
+    private function getCsrfToken(ServerRequestInterface $request): ?string
+    {
+        if ($this->enableCookie) {
+            $cookies = $request->getCookieParams();
+            return $cookies[$this->name] ?? null;
+        } else {
+            return $this->session->get($this->name);
+        }
+    }
+
+    private function validateCsrfToken(ServerRequestInterface $request, ResponseInterface $response): ResponseInterface
+    {
+        $trueToken = $this->getCsrfToken($request);
+        $method = $request->getMethod();
+
+        if (\in_array($method, ['GET', 'HEAD', 'OPTIONS'], true)) {
+            return $response;
+        }
+
+        $unmaskedToken = $this->getTokenFromRequest($request);
+        if (empty($unmaskedToken) || !hash_equals($unmaskedToken, $trueToken)) {
+            $response = $this->responseFactory->createResponse(400);
```

Comment:
Not sure if status code should be 400 or 403

Reply:
400 is fine.